File: backend/terry/detect_exe.py
#!/usr/bin/env python3
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# Copyright 2017-2018 - Edoardo Morassutto <edoardo.morassutto@gmail.com>

# Inspired by
# http://lallouslab.net/2017/07/03/detect-executable-format-using-python/

EXEFLAG_NONE = 0x0000
EXEFLAG_LINUX = 0x0001
EXEFLAG_WINDOWS = 0x0002
EXEFLAG_MACOS = 0x0004
EXEFLAG_MACOS_FAT = 0x0008
EXEFLAG_32BITS = 0x0010
EXEFLAG_64BITS = 0x0020

# Keep signatures sorted by size
_EXE_SIGNATURES = (
    (b"\x4D\x5A", EXEFLAG_WINDOWS),
    (b"\xCE\xFA\xED\xFE", EXEFLAG_MACOS | EXEFLAG_32BITS),
    (b"\xCF\xFA\xED\xFE", EXEFLAG_MACOS | EXEFLAG_64BITS),
    (b"\xBE\xBA\xFE\xCA", EXEFLAG_MACOS | EXEFLAG_32BITS | EXEFLAG_MACOS_FAT),
    (b"\xBF\xBA\xFE\xCA", EXEFLAG_MACOS | EXEFLAG_64BITS | EXEFLAG_MACOS_FAT),
    (b"\x7F\x45\x4C\x46\x01", EXEFLAG_LINUX | EXEFLAG_32BITS),
    (b"\x7F\x45\x4C\x46\x02", EXEFLAG_LINUX | EXEFLAG_64BITS),
)


def get_exeflags(content: bytes) -> int:
    buf = content[0:5]
    for sig, flags in _EXE_SIGNATURES:
        sig_len = len(sig)
        if buf[0:sig_len] == sig:
            return flags
    return EXEFLAG_NONE

File: backend/src/detect_exe.rs
use bitflags::bitflags;

bitflags! {
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    pub struct ExeFlags: u32 {
        const NONE      = 0x0000;
        const LINUX     = 0x0001;
        const WINDOWS   = 0x0002;
        const MACOS     = 0x0004;
        const MACOS_FAT = 0x0008;
        const BITS_32   = 0x0010;
        const BITS_64   = 0x0020;
    }
}

// Keep signatures sorted by size (longer ones first) to ensure correct matching.
const EXE_SIGNATURES: &[(&[u8], ExeFlags)] = &[
    // 5-byte signatures
    (
        b"\x7F\x45\x4C\x46\x01",
        ExeFlags::LINUX.union(ExeFlags::BITS_32),
    ),
    (
        b"\x7F\x45\x4C\x46\x02",
        ExeFlags::LINUX.union(ExeFlags::BITS_64),
    ),
    // 4-byte signatures
    (
        b"\xCE\xFA\xED\xFE",
        ExeFlags::MACOS.union(ExeFlags::BITS_32),
    ),
    (
        b"\xCF\xFA\xED\xFE",
        ExeFlags::MACOS.union(ExeFlags::BITS_64),
    ),
    (
        b"\xBE\xBA\xFE\xCA",
        ExeFlags::MACOS
            .union(ExeFlags::BITS_32)
            .union(ExeFlags::MACOS_FAT),
    ),
    (
        b"\xBF\xBA\xFE\xCA",
        ExeFlags::MACOS
            .union(ExeFlags::BITS_64)
            .union(ExeFlags::MACOS_FAT),
    ),
    // 2-byte signature
    (b"\x4D\x5A", ExeFlags::WINDOWS),
];

#[allow(dead_code)]
pub fn get_exeflags(content: &[u8]) -> ExeFlags {
    for (sig, flags) in EXE_SIGNATURES {
        if content.starts_with(sig) {
            return *flags;
        }
    }
    ExeFlags::NONE
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_linux_64() {
        let content = b"\x7F\x45\x4C\x46\x02\x01\x01\x00";
        assert_eq!(get_exeflags(content), ExeFlags::LINUX | ExeFlags::BITS_64);
    }

    #[test]
    fn test_linux_32() {
        let content = b"\x7F\x45\x4C\x46\x01\x01\x01\x00";
        assert_eq!(get_exeflags(content), ExeFlags::LINUX | ExeFlags::BITS_32);
    }

    #[test]
    fn test_windows() {
        let content = b"\x4D\x5A\x90\x00";
        assert_eq!(get_exeflags(content), ExeFlags::WINDOWS);
    }

    #[test]
    fn test_macos_64() {
        let content = b"\xCF\xFA\xED\xFE\x07\x00\x00\x01";
        assert_eq!(get_exeflags(content), ExeFlags::MACOS | ExeFlags::BITS_64);
    }

    #[test]
    fn test_macos_32() {
        let content = b"\xCE\xFA\xED\xFE\x07\x00\x00\x00";
        assert_eq!(get_exeflags(content), ExeFlags::MACOS | ExeFlags::BITS_32);
    }

    #[test]
    fn test_macos_fat_64() {
        let content = b"\xBF\xBA\xFE\xCA";
        assert_eq!(
            get_exeflags(content),
            ExeFlags::MACOS | ExeFlags::BITS_64 | ExeFlags::MACOS_FAT
        );
    }

    #[test]
    fn test_macos_fat_32() {
        let content = b"\xBE\xBA\xFE\xCA";
        assert_eq!(
            get_exeflags(content),
            ExeFlags::MACOS | ExeFlags::BITS_32 | ExeFlags::MACOS_FAT
        );
    }

    #[test]
    fn test_none() {
        let content = b"this is not an executable";
        assert_eq!(get_exeflags(content), ExeFlags::NONE);
    }

    #[test]
    fn test_empty() {
        let content = b"";
        assert_eq!(get_exeflags(content), ExeFlags::NONE);
    }
}
